peribolos: Add capability to update existing repos

petr-muller opened the pull request:
A continuation of https://github.com/kubernetes/test-infra/pull/14732. Adds a capability to update the fields already supported when creating, plus two fields that can only be set on update (not on creation): `archived` and `default_branch`. A special case of GH API not supporting un-archival of repos is handled locally.

The feature notably does not support repo renaming yet (or more precisely, the GH client change does, but peribolos change does not).

/cc @stevekuznetsov @fejta

fejta commented:
For renaming, I like the previously pattern: https://github.com/kubernetes/test-infra/tree/master/label_sync#configuration

petr-muller commented:
@stevekuznetsov Good point. I think (for both publishing and archiving) the explicit allowing options are more appropriate. For publishing, even a single mistakenly published private repo is a potential problem. For both archival and publishing, it is hard to imagine how someone would by mistake affect many repos: each repo would need an explicit config line, it is not possible to set these e.g. org-wide. So I think just an option for confirming that you want to allow these problematic operations is fine.

Ad renames, the `previously` pattern is great, no reason for deviating from that.

petr-muller commented:
Setting to WIP for adding the explicit allowing options for archival and publishing.

petr-muller commented:
Updated: Added two new parameter to allow repo archivals and publishing.

Additionally, I have changed how peribolos behaves when these disallowed changes occur. Previously, such change would cause the given repo to not be updated at all, even if there were changes that could be made otherwise. Now, the disallowed change is just removed from the delta used for update, so the remaining ("allowed") changes can still be done. This was done in a separate commit, so it can be easily removed from this PR, if desired.

petr-muller commented on prow/cmd/peribolos/main.go:
```diff
@@ -908,17 +967,38 @@ func configureRepos(client repoClient, orgName string, orgConfig org.Config) err
 			logrus.WithField("repo", wantName).Info("repo does not exist, creating")
-			if _, err := client.CreateRepo(orgName, false, newRepoCreateRequest(wantName, wantRepo)); err != nil {
-				createErrors = append(createErrors, err)
+			created, err := client.CreateRepo(orgName, false, newRepoCreateRequest(wantName, wantRepo))
```
Correct. Or set a `default_branch` on it, which unfortunately fails, because the repo may not have any git content yet. It is not ideal.

The question is, how to handle this situation better? If peribolos won't do these `UpdateRepo` calls in the same run, it will do them in the next run anyway (I assume peribolos is intended to be run regularly & often). Maybe it should error out: I usually prefer tools to not be opinionated about what is supposed to be a valid operation, but if desired, I can add check for this.

fejta replied:
I would say that if the repo is to be archived and doesn't yet exist then don't create it

stevekuznetsov commented:
I'm ok tackling the archival errors later if we end up hitting that edge.

/hold cancel
